python/pyspark/util.py

```diff
@@ -263,6 +264,69 @@ def _parse_memory(s):
     return int(float(s[:-1]) * units[s[-1].lower()])
 
 
+def is_pinned_thread_mode():
+    """
+    Return ``True`` when spark run under pinned thread mode.
+    """
+    from pyspark import SparkContext
+    return isinstance(SparkContext._gateway, ClientServer)
+
+
+def inheritable_thread_target(f):
+    """
+    Return thread target wrapper which is recommended to be used in PySpark when the
+    pinned thread mode is enabled. The wrapper function, before calling original
+    thread target, it inherits the inheritable properties specific
+    to JVM thread such as ``InheritableThreadLocal``.
+
+    Also, note that pinned thread mode does not close the connection from Python
+    to JVM when the thread is finished in the Python side. With this wrapper, Python
+    garbage-collects the Python thread instance and also closes the connection
+    which finishes JVM thread correctly.
+
+    When the pinned thread mode is off, it return the original ``f``.
+    :param f: the original thread target.
+
+    .. versionadded:: 3.1.0
```

Comment:
I think you meant: 3.2.0

Reply:
i think we need backport to 3.1

Reply:
I'd also not backport unless it's essential - how bad is it?

Reply:
Ok we don't need backport it in Apache/spark.